src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+
+	rc = agg_data_extent(entry, agg_entry, ih, acts);
+
+	return rc;
+}
+
+/* Pre-subtree iteration call back for per-object iterator
+ */
+static int
+agg_iterate_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+	vos_iter_type_t type, vos_iter_param_t *param,
+	void *cb_arg, unsigned int *acts)
+{
+	struct ec_agg_entry	*agg_entry = (struct ec_agg_entry *) cb_arg;
+	int			 rc = 0;
+
+	switch (type) {
+	case VOS_ITER_DKEY:
+		rc = agg_dkey(ih, entry, agg_entry, acts);
+		break;
+	case VOS_ITER_AKEY:
+		rc = agg_akey(ih, entry, agg_entry, acts);
+		break;
+	case VOS_ITER_RECX:
+		rc = agg_ev(ih, entry, agg_entry, acts);
+		break;
+	default:
+		break;
+	}
+
+	return rc;
+}
+
+/* Post iteration call back for per-object iterator
+ */
+static int
+agg_iterate_post_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+	vos_iter_type_t type, vos_iter_param_t *param,
+	void *cb_arg, unsigned int *acts)
+{
+	struct ec_agg_entry	*agg_entry = (struct ec_agg_entry *) cb_arg;
+	int			 rc = 0;
+
+	switch (type) {
+	case VOS_ITER_DKEY:
+		break;
+	case VOS_ITER_AKEY:
+		rc = agg_akey_post(ih, entry, agg_entry, acts);
+		break;
+	case VOS_ITER_RECX:
+		break;
+	default:
+		break;
+	}
+
+	return rc;
+}
+
+/* Initializes the struct holding the iteration state (ec_agg_entry).
+ */
+static void
+agg_init_entry(daos_handle_t lcoh, daos_handle_t gcoh,
+	       struct ec_agg_entry *agg_entry,
+	       vos_iter_entry_t *entry, struct daos_oclass_attr *oca,
+	       d_sg_list_t *sgl)
+{
+	agg_entry->ae_oid		= entry->ie_oid;
+	agg_entry->ae_oca		= oca;
+	agg_entry->ae_sgl		= sgl;
+	agg_entry->ae_chdl		= lcoh;
+	agg_entry->ae_cont_hdl		= gcoh;
+	agg_entry->ae_rsize		= 0UL;
+	agg_entry->ae_obj_hdl		= DAOS_HDL_INVAL;
+
+	memset(&agg_entry->ae_dkey, 0, sizeof(agg_entry->ae_dkey));
+	memset(&agg_entry->ae_akey, 0, sizeof(agg_entry->ae_akey));
+	memset(&agg_entry->ae_par_extent, 0, sizeof(agg_entry->ae_par_extent));
+
+	agg_entry->ae_cur_stripe.as_stripenum	= 0UL;
+	agg_entry->ae_cur_stripe.as_hi_epoch	= 0UL;
+	agg_entry->ae_cur_stripe.as_stripe_fill = 0UL;
+	agg_entry->ae_cur_stripe.as_extent_cnt	= 0U;
+}
+
+/* Configures and invokes nested iterator. Called from full-VOS object
+ * iteration.
+ */
+static int
+agg_subtree_iterate(daos_handle_t ih, daos_unit_oid_t *oid,
+		    struct ec_agg_param *agg_param)
+{
+	vos_iter_param_t	 iter_param = { 0 };
+	struct vos_iter_anchors  anchors = { 0 };
+	int			 rc = 0;
+
+	iter_param.ip_hdl		= DAOS_HDL_INVAL;
+	iter_param.ip_ih		= ih;
+	iter_param.ip_flags		= VOS_IT_RECX_VISIBLE;
+	iter_param.ip_oid		= *oid;
+	iter_param.ip_epr.epr_lo	= 0ULL;
+	iter_param.ip_epr.epr_hi	= DAOS_EPOCH_MAX;
+	iter_param.ip_epc_expr		= VOS_IT_EPC_RR;
+	iter_param.ip_flags		= VOS_IT_RECX_VISIBLE;
+	iter_param.ip_recx.rx_idx	= 0ULL;
+	iter_param.ip_recx.rx_nr	= ~PARITY_INDICATOR;
+
+	rc = vos_iterate(&iter_param, VOS_ITER_DKEY, true, &anchors,
+			 agg_iterate_pre_cb, agg_iterate_post_cb,
+			 agg_param->ap_agg_entry, NULL);
```

Comment:
Why don't just directly call recursive vos_iterrate() from OBJ level? I don't quite see why it's called in the way of "vos_iterate(OBJ, non-recursive) -> callback() -> vos_iterate(DKEY, recursive)"

The problem of this way is that you have no idea if the vos_iterate(DKEY, recursvie) yield, so you don't know if iteration in OBJ level needs re-probe.

Reply:
The non-recursive iterator at the object level is used by the iterate_all method to find eligible OIDs (EC and this target is leader).  No need for recursive exploration unless OID is eligible.  There is another method (based on committed DTX entries) that establishes eligible OIDs outside of VOS iteration (I was told to remove this code for the first patch).